pkg/mcs/resourcemanager/server/keyspace_manager.go

```diff
@@ -346,3 +347,185 @@ func (rt *ruTracker) getRUPerSec() float64 {
+	serviceLimiter := krgm.getServiceLimiter()
+	// No need to conciliate if the service limit is not set or is 0.
+	if serviceLimiter == nil || serviceLimiter.ServiceLimit == 0 {
+		return
+	}
+	priorityQueues := krgm.getPriorityQueues()
+	if len(priorityQueues) == 0 {
+		return
+	}
+	remainingServiceLimit := serviceLimiter.ServiceLimit
+	for _, queue := range priorityQueues {
+		if len(queue) == 0 {
+			continue
+		}
+		// If the remaining service limit is 0, set the fill rate of all the remaining resource groups to 0 directly.
+		// Only apply this to active resource groups (those with RU tracker and actual RU consumption).
+		if remainingServiceLimit == 0 {
+			for _, group := range queue {
+				ruTracker := krgm.getRUTracker(group.Name)
+				if ruTracker == nil || ruTracker.getRUPerSec() == 0 {
+					continue
+				}
+				// Only set the active resource groups to avoid unnecessary overrides.
+				group.overrideFillRateAndBurstLimit(0, 0)
+			}
+			continue
+		}
+
+		var (
+			totalBasicRUDemand = 0.0
+			totalBurstRUDemand = 0.0
+			basicRUDemandMap   = make(map[string]float64, len(queue))
+			burstRUDemandMap   = make(map[string]float64, len(queue))
+		)
+		// Calculate the basic and burst RU demands for all groups in this priority level.
+		for _, group := range queue {
+			ruTracker := krgm.getRUTracker(group.Name)
+			// Not found the RU tracker, skip this group.
+			if ruTracker == nil {
+				continue
+			}
+			ruPerSec := ruTracker.getRUPerSec()
+			fillRateSetting := group.getFillRateSetting()
+			burstLimitSetting := group.getBurstLimitSetting()
+			// Calculate the basic RU demand of the resource group.
+			// Basic demand is the minimum of real-time RU consumption and configured fill rate.
+			basicRUDemand := math.Min(ruPerSec, fillRateSetting)
+			totalBasicRUDemand += basicRUDemand
+			basicRUDemandMap[group.Name] = basicRUDemand
+			// Calculate the burst RU demand of the resource group.
+			// Burst demand is the excess consumption beyond the fill rate setting.
+			burstRUDemand := math.Max(0, ruPerSec-fillRateSetting)
+			// If the burst limit setting is greater than 0, the burst RU demand should not exceed the burst limit setting.
+			if burstLimitSetting >= 0 {
+				burstRUDemand = math.Min(burstRUDemand, burstLimitSetting)
+			}
+			totalBurstRUDemand += burstRUDemand
+			burstRUDemandMap[group.Name] = burstRUDemand
+		}
+
+		totalRUDemand := totalBasicRUDemand + totalBurstRUDemand
+		// If the total real-time RU demand is greater than or equal to the remaining service limit,
+		// we need to divide the remaining service limit proportionally within the same priority level.
+		if totalRUDemand >= remainingServiceLimit {
+			// If the basic RU demand already exceeds the remaining service limit, then we can only try to meet the basic needs of
+			// the resource groups through proportional allocation of the fill rate setting.
+			if totalBasicRUDemand >= remainingServiceLimit {
+				for _, group := range queue {
```

Comment:
Here you may need first sort the groups by normalizedRUDemand(RUDemand/FillRate) and handle groups by this order. So if some groups' normalizedRUDemand is smaller than totalBasicRUDemand/totalBasicRUDemand, that means it's required ru is smaller than its proportion, then we can retain this unused RUs to other needed groups.
Example,  say total limit is 1000 there are 3 groups with (ru_per_sec, ru_required) values (1000, 100), (1000, 400), (1000, 800). The ideal granted override limit should be 100, 400, 500. But in the algorithm in this PR, the override limit will be 100, 333, 333.